File: src/types/gacha.ts
// Gacha history types for HoYoverse games

export type GachaGame = 'genshin' | 'star_rail' | 'zzz';

export interface GachaType {
  id: string;
  name: string;
}

export interface GachaRecord {
  id: string;
  uid: string;
  gacha_type: string;
  item_id: string | null;
  name: string;
  item_type: string;
  rank_type: string;
  time: string;
}

export interface GachaHistory {
  game: GachaGame;
  uid: string;
  records: GachaRecord[];
  last_sync: number;
  region: string | null;
}

export interface FiveStarPull {
  name: string;
  pity: number;
  time: string;
}

export interface BannerStats {
  total_pulls: number;
  five_star_count: number;
  four_star_count: number;
  current_pity: number;
  average_pity: number;
  five_star_pulls: FiveStarPull[];
}

export interface GachaStats {
  total_pulls: number;
  five_star_count: number;
  four_star_count: number;
  three_star_count: number;
  banner_stats: Record<string, BannerStats>;
}

export interface GachaAccount {
  game: GachaGame;
  uid: string;
  last_sync: number;
  total_records: number;
  region: string | null;
}

export interface DetectedGachaGame {
  game: GachaGame;
  install_path: string;
  cache_exists: boolean;
  icon_path: string | null;
}

export interface RefreshGachaRequest {
  game: GachaGame;
  game_path: string;
}

// UIGF v4 export format
export interface UigfInfo {
  export_timestamp: number;
  export_app: string;
  export_app_version: string;
  version: string;
}

export interface UigfRecord {
  id: string;
  gacha_type: string;
  item_id?: string;
  name: string;
  item_type: string;
  rank_type: string;
  time: string;
}

export interface UigfGameData {
  uid: string;
  list: UigfRecord[];
}

export interface UigfExport {
  info: UigfInfo;
  hk4e?: UigfGameData[];
  hkrpg?: UigfGameData[];
  nap?: UigfGameData[];
}

// Helper functions

export function getGameDisplayName(game: GachaGame): string {
  switch (game) {
    case 'genshin':
      return 'Genshin Impact';
    case 'star_rail':
      return 'Honkai: Star Rail';
    case 'zzz':
      return 'Zenless Zone Zero';
    default:
      return 'Unknown';
  }
}

export function getGameShortName(game: GachaGame): string {
  switch (game) {
    case 'genshin':
      return 'GI';
    case 'star_rail':
      return 'HSR';
    case 'zzz':
      return 'ZZZ';
    default:
      return '?';
  }
}

export function getGameIconFilename(game: GachaGame): string {
  switch (game) {
    case 'genshin':
      return 'GenshinImpact.png';
    case 'star_rail':
      return 'StarRail.png';
    case 'zzz':
      return 'ZenlessZoneZero.png';
  }
}

export function getGachaTypeName(game: GachaGame, gachaType: string): string {
  const typeNames: Record<GachaGame, Record<string, string>> = {
    genshin: {
      '301': 'Character Event',
      '302': 'Weapon Event',
      '200': 'Standard',
      '100': 'Beginner',
      '500': 'Chronicled Wish',
    },
    star_rail: {
      '11': 'Character Event',
      '12': 'Light Cone Event',
      '1': 'Standard',
      '2': 'Departure',
    },
    zzz: {
      '2001': 'Exclusive Channel',
      '3001': 'W-Engine Channel',
      '1001': 'Standard Channel',
      '5001': 'Bangboo Channel',
    },
  };

  return typeNames[game]?.[gachaType] || `Banner ${gachaType}`;
}

export function getGachaTypes(game: GachaGame): GachaType[] {
  const types: Record<GachaGame, GachaType[]> = {
    genshin: [
      { id: '301', name: 'Character Event' },
      { id: '302', name: 'Weapon Event' },
      { id: '200', name: 'Standard' },
      { id: '100', name: 'Beginner' },
      { id: '500', name: 'Chronicled Wish' },
    ],
    star_rail: [
      { id: '11', name: 'Character Event' },
      { id: '12', name: 'Light Cone Event' },
      { id: '1', name: 'Standard' },
      { id: '2', name: 'Departure' },
    ],
    zzz: [
      { id: '2001', name: 'Exclusive Channel' },
      { id: '3001', name: 'W-Engine Channel' },
      { id: '1001', name: 'Standard Channel' },
      { id: '5001', name: 'Bangboo Channel' },
    ],
  };

  return types[game] || [];
}

export function getRarityColor(rarity: string | number): string {
  const r = typeof rarity === 'string' ? parseInt(rarity) : rarity;
  switch (r) {
    case 5:
      return 'text-amber-400';
    case 4:
      return 'text-purple-400';
    default:
      return 'text-blue-400';
  }
}

export function getRarityBgColor(rarity: string | number): string {
  const r = typeof rarity === 'string' ? parseInt(rarity) : rarity;
  switch (r) {
    case 5:
      return 'bg-amber-400/20';
    case 4:
      return 'bg-purple-400/20';
    default:
      return 'bg-blue-400/20';
  }
}

export function formatTimestamp(ms: number): string {
  if (!ms) return 'Never';
  const date = new Date(ms);
  return date.toLocaleString();
}

export function formatRelativeTime(ms: number): string {
  if (!ms) return 'Never';

  const now = Date.now();
  const diff = now - ms;

  const seconds = Math.floor(diff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);

  if (days > 0) return `${days}d ago`;
  if (hours > 0) return `${hours}h ago`;
  if (minutes > 0) return `${minutes}m ago`;
  return 'Just now';
}

export function calculateFiveStarRate(stats: GachaStats): number {
  if (stats.total_pulls === 0) return 0;
  return (stats.five_star_count / stats.total_pulls) * 100;
}

export function calculateFourStarRate(stats: GachaStats): number {
  if (stats.total_pulls === 0) return 0;
  return (stats.four_star_count / stats.total_pulls) * 100;
}

// Soft pity thresholds
export function getSoftPityThreshold(game: GachaGame, gachaType: string): number {
  // Most banners have soft pity at around 74-75 for 5-star
  // Weapon banners typically have lower pity
  if (game === 'genshin' && gachaType === '302') {
    return 63; // Weapon banner
  }
  return 74; // Standard for most banners
}

export function getHardPityThreshold(game: GachaGame, gachaType: string): number {
  if (game === 'genshin' && gachaType === '302') {
    return 80; // Weapon banner
  }
  return 90; // Standard for most banners
}

File: src-tauri/src/models/gacha.rs
// Gacha history models for HoYoverse games (Genshin Impact, Honkai Star Rail, ZZZ)

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Supported HoYoverse games for gacha history
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GachaGame {
    Genshin,
    StarRail,
    Zzz,
}

impl GachaGame {
    pub fn display_name(&self) -> &'static str {
        match self {
            GachaGame::Genshin => "Genshin Impact",
            GachaGame::StarRail => "Honkai: Star Rail",
            GachaGame::Zzz => "Zenless Zone Zero",
        }
    }

    pub fn short_name(&self) -> &'static str {
        match self {
            GachaGame::Genshin => "genshin",
            GachaGame::StarRail => "starrail",
            GachaGame::Zzz => "zzz",
        }
    }

    /// UIGF v4 key for this game
    pub fn uigf_key(&self) -> &'static str {
        match self {
            GachaGame::Genshin => "hk4e",
            GachaGame::StarRail => "hkrpg",
            GachaGame::Zzz => "nap",
        }
    }

    /// Get web cache path relative to game install directory
    pub fn cache_path(&self) -> &'static str {
        match self {
            GachaGame::Genshin => "GenshinImpact_Data/webCaches/Cache/Cache_Data/data_2",
            GachaGame::StarRail => "StarRail_Data/webCaches/Cache/Cache_Data/data_2",
            GachaGame::Zzz => "ZenlessZoneZero_Data/webCaches/Cache/Cache_Data/data_2",
        }
    }

    /// API endpoint for gacha history (Global version)
    pub fn api_endpoint(&self) -> &'static str {
        match self {
            GachaGame::Genshin => "https://public-operation-hk4e-sg.hoyoverse.com/gacha_info/api/getGachaLog",
            GachaGame::StarRail => "https://public-operation-hkrpg-sg.hoyoverse.com/common/gacha_record/api/getGachaLog",
            GachaGame::Zzz => "https://public-operation-nap-sg.hoyoverse.com/common/gacha_record/api/getGachaLog",
        }
    }

    /// Get gacha/banner types for this game
    pub fn gacha_types(&self) -> Vec<GachaType> {
        match self {
            GachaGame::Genshin => vec![
                GachaType { id: "301".to_string(), name: "Character Event".to_string() },
                GachaType { id: "302".to_string(), name: "Weapon Event".to_string() },
                GachaType { id: "200".to_string(), name: "Standard".to_string() },
                GachaType { id: "100".to_string(), name: "Beginner".to_string() },
                GachaType { id: "500".to_string(), name: "Chronicled Wish".to_string() },
            ],
            GachaGame::StarRail => vec![
                GachaType { id: "11".to_string(), name: "Character Event".to_string() },
                GachaType { id: "12".to_string(), name: "Light Cone Event".to_string() },
                GachaType { id: "1".to_string(), name: "Standard".to_string() },
                GachaType { id: "2".to_string(), name: "Departure".to_string() },
            ],
            GachaGame::Zzz => vec![
                GachaType { id: "2001".to_string(), name: "Exclusive Channel".to_string() },
                GachaType { id: "3001".to_string(), name: "W-Engine Channel".to_string() },
                GachaType { id: "1001".to_string(), name: "Standard Channel".to_string() },
                GachaType { id: "5001".to_string(), name: "Bangboo Channel".to_string() },
            ],
        }
    }
}

/// Banner/gacha type information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GachaType {
    pub id: String,
    pub name: String,
}

/// A single gacha/wish record
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GachaRecord {
    /// Unique ID from the API
    pub id: String,
    /// User ID
    pub uid: String,
    /// Banner/gacha type ID
    pub gacha_type: String,
    /// Item ID (if available)
    pub item_id: Option<String>,
    /// Item name
    pub name: String,
    /// Item type (Character/Weapon/Light Cone/W-Engine/Bangboo)
    pub item_type: String,
    /// Rarity (3, 4, or 5)
    pub rank_type: String,
    /// Pull timestamp (ISO 8601 format)
    pub time: String,
}

impl GachaRecord {
    /// Get rarity as a number
    pub fn rarity(&self) -> u8 {
        self.rank_type.parse().unwrap_or(3)
    }
}

/// Gacha history for a specific account
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GachaHistory {
    /// Game this history belongs to
    pub game: GachaGame,
    /// User ID
    pub uid: String,
    /// All gacha records, sorted by ID (newest first)
    pub records: Vec<GachaRecord>,
    /// Last sync timestamp (Unix ms)
    pub last_sync: u64,
    /// Region/server (e.g., "os_asia", "os_euro", "os_usa")
    pub region: Option<String>,
}

impl GachaHistory {
    pub fn new(game: GachaGame, uid: String) -> Self {
        Self {
            game,
            uid,
            records: Vec::new(),
            last_sync: 0,
            region: None,
        }
    }

    /// Merge new records into existing history (incremental sync)
    /// Returns the number of new records added
    pub fn merge(&mut self, new_records: Vec<GachaRecord>) -> usize {
        let existing_ids: std::collections::HashSet<String> =
            self.records.iter().map(|r| r.id.clone()).collect();

        let mut added = 0;
        for record in new_records {
            if !existing_ids.contains(&record.id) {
                self.records.push(record);
                added += 1;
            }
        }

        // Sort by ID descending (newest first)
        self.records.sort_by(|a, b| b.id.cmp(&a.id));
        self.last_sync = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        added
    }

    /// Get the latest record ID for incremental sync
    pub fn latest_id(&self) -> Option<&str> {
        self.records.first().map(|r| r.id.as_str())
    }

    /// Calculate statistics for this history
    pub fn calculate_stats(&self) -> GachaStats {
        let mut stats = GachaStats::default();
        let mut banner_records: HashMap<String, Vec<&GachaRecord>> = HashMap::new();

        for record in &self.records {
            banner_records
                .entry(record.gacha_type.clone())
                .or_default()
                .push(record);
        }

        stats.total_pulls = self.records.len();
        stats.five_star_count = self.records.iter().filter(|r| r.rarity() == 5).count();
        stats.four_star_count = self.records.iter().filter(|r| r.rarity() == 4).count();
        stats.three_star_count = self.records.iter().filter(|r| r.rarity() == 3).count();

        // Calculate per-banner stats
        for (gacha_type, records) in banner_records {
            let banner_stats = Self::calculate_banner_stats(&records);
            stats.banner_stats.insert(gacha_type, banner_stats);
        }

        stats
    }

    fn calculate_banner_stats(records: &[&GachaRecord]) -> BannerStats {
        let mut stats = BannerStats::default();
        stats.total_pulls = records.len();

        // Records are sorted newest first, so we need to reverse for pity calculation
        let mut pity_counter = 0;
        let mut pity_history: Vec<u32> = Vec::new();

        for record in records.iter().rev() {
            pity_counter += 1;
            if record.rarity() == 5 {
                pity_history.push(pity_counter);
                stats.five_star_pulls.push(FiveStarPull {
                    name: record.name.clone(),
                    pity: pity_counter,
                    time: record.time.clone(),
                });
                pity_counter = 0;
            }
        }

        stats.current_pity = pity_counter;
        stats.five_star_count = pity_history.len();
        stats.four_star_count = records.iter().filter(|r| r.rarity() == 4).count();

        if !pity_history.is_empty() {
            stats.average_pity = pity_history.iter().sum::<u32>() as f64 / pity_history.len() as f64;
        }

        stats
    }
}

/// Aggregated gacha statistics
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct GachaStats {
    pub total_pulls: usize,
    pub five_star_count: usize,
    pub four_star_count: usize,
    pub three_star_count: usize,
    /// Stats per banner type
    pub banner_stats: HashMap<String, BannerStats>,
}

impl GachaStats {
    pub fn five_star_rate(&self) -> f64 {
        if self.total_pulls == 0 {
            return 0.0;
        }
        (self.five_star_count as f64 / self.total_pulls as f64) * 100.0
    }

    pub fn four_star_rate(&self) -> f64 {
        if self.total_pulls == 0 {
            return 0.0;
        }
        (self.four_star_count as f64 / self.total_pulls as f64) * 100.0
    }
}

/// Statistics for a specific banner
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct BannerStats {
    pub total_pulls: usize,
    pub five_star_count: usize,
    pub four_star_count: usize,
    pub current_pity: u32,
    pub average_pity: f64,
    /// List of 5-star pulls with pity info
    pub five_star_pulls: Vec<FiveStarPull>,
}

/// Information about a 5-star pull
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FiveStarPull {
    pub name: String,
    pub pity: u32,
    pub time: String,
}

/// Account info for gacha history
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GachaAccount {
    pub game: GachaGame,
    pub uid: String,
    pub last_sync: u64,
    pub total_records: usize,
    pub region: Option<String>,
}

/// Request to refresh gacha history
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RefreshGachaRequest {
    pub game: GachaGame,
    pub game_path: String,
}

/// Result from Python worker
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GachaWorkerResult {
    pub uid: String,
    pub records: Vec<GachaRecord>,
    pub region: Option<String>,
}

/// UIGF v4 export format
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UigfExport {
    pub info: UigfInfo,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hk4e: Option<Vec<UigfGameData>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hkrpg: Option<Vec<UigfGameData>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nap: Option<Vec<UigfGameData>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UigfInfo {
    pub export_timestamp: u64,
    pub export_app: String,
    pub export_app_version: String,
    pub version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UigfGameData {
    pub uid: String,
    pub list: Vec<UigfRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UigfRecord {
    pub id: String,
    pub gacha_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub item_id: Option<String>,
    pub name: String,
    pub item_type: String,
    pub rank_type: String,
    pub time: String,
}

impl From<&GachaRecord> for UigfRecord {
    fn from(record: &GachaRecord) -> Self {
        Self {
            id: record.id.clone(),
            gacha_type: record.gacha_type.clone(),
            item_id: record.item_id.clone(),
            name: record.name.clone(),
            item_type: record.item_type.clone(),
            rank_type: record.rank_type.clone(),
            time: record.time.clone(),
        }
    }
}

/// Detected game installation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectedGachaGame {
    pub game: GachaGame,
    pub install_path: String,
    pub cache_exists: bool,
    pub icon_path: Option<String>,
}
